src/repr/src/relation.rs

```diff
@@ -367,6 +367,21 @@ pub struct ColumnIndex(usize);
 
 static_assertions::assert_not_impl_all!(ColumnIndex: Arbitrary);
 
+impl ColumnIndex {
```

Comment:
Good callout, I added some docs the describe the relationship between `RelationDesc`, `ColumnMetadata`, and `RelationType`, let me know if it makes sense or requires more detail!

I'm pretty sure `RelationDesc::iter` is still correct since the order of columns in the `RelationType` and `metadata` will be the same, but that's just by chance. I updated the impl to read from `metadata` to determine the index we should read from in the `RelationType`.